core/router/gateway/gateway.go

```diff
@@ -26,4 +31,71 @@ type Gateway struct {
 	Utilization Utilization
 	Schedule    Schedule
 	LastSeen    time.Time
+	Token       string
+
+	monitor *monitorConn
+
+	Ctx log.Interface
+}
+
+func (g *Gateway) updateTimestamp() {
+	g.LastSeen = time.Now()
+}
+
+func (g *Gateway) HandleStatus(status *pb.Status) (err error) {
+	if g.monitor != nil {
+		go func() {
+			cl, err := g.statusMonitor()
+			if err != nil {
+				g.Ctx.WithError(errors.FromGRPCError(err)).Error("Failed to establish status connection to the NOC")
+			}
+
+			if err = cl.Send(status); err != nil {
+				g.Ctx.WithError(errors.FromGRPCError(err)).Error("NOC status push failed")
+			}
+		}()
+	}
+
+	g.updateTimestamp()
+	return g.Status.Update(status)
+}
+
+func (g *Gateway) HandleUplink(uplink *pb_router.UplinkMessage) (err error) {
+	if g.monitor != nil {
+		go func() {
+			cl, err := g.uplinkMonitor()
+			if err != nil {
+				g.Ctx.WithError(errors.FromGRPCError(err)).Error("Failed to establish uplink connection to the NOC")
+			}
+
+			if err = cl.Send(uplink); err != nil {
+				g.Ctx.WithError(errors.FromGRPCError(err)).Error("NOC uplink push failed")
+			}
+		}()
+	}
+
+	g.updateTimestamp()
+	g.Schedule.Sync(uplink.GatewayMetadata.Timestamp)
+	return g.Utilization.AddRx(uplink)
+}
+
+func (g *Gateway) HandleDownlink(identifier string, downlink *pb_router.DownlinkMessage) (err error) {
+	if g.monitor != nil {
```

Comment:
We probably want to do this after we get the result, because at this point the func might still fail